sound/soc/sof/debug.c

```diff
@@ -17,6 +17,244 @@
+#define MAX_IPC_FLOOD_DURATION_MS 1000
+#define MAX_IPC_FLOOD_COUNT 10000
+#define IPC_FLOOD_TEST_RESULT_LEN 512
+
+static int sof_debug_ipc_flood_test(struct snd_sof_dev *sdev,
+				    struct snd_sof_dfsentry *dfse,
+				    bool flood_duration_test,
+				    unsigned long ipc_duration_ms,
+				    unsigned long ipc_count)
+{
+	struct sof_ipc_cmd_hdr hdr;
+	struct sof_ipc_reply reply;
+	char *output4, *output5, *output6;
+	u64 min_response_time = U64_MAX;
+	ktime_t start, end, test_end;
+	u64 avg_response_time = 0;
+	u64 max_response_time = 0;
+	char  *output2, *output3;
+	u64 ipc_response_time;
+	char *output1 = NULL;
+	int i = 0;
+	int size;
+	int ret;
+
+	/* configure test IPC */
+	hdr.cmd = SOF_IPC_GLB_TEST | SOF_IPC_TEST_IPC_FLOOD;
+	hdr.size = sizeof(hdr);
+
+	/* set test end time for duration flood test */
+	if (flood_duration_test)
+		test_end = ktime_get_ns() + ipc_duration_ms * NSEC_PER_MSEC;
+
+	/* send test IPC's */
+	while (1) {
+		start = ktime_get();
+		ret = sof_ipc_tx_message(sdev->ipc, hdr.cmd, &hdr, hdr.size,
+					 &reply, sizeof(reply));
+		end = ktime_get();
+
+		if (ret < 0)
+			break;
+
+		/* compute min and max response times */
+		ipc_response_time = ktime_to_ns(ktime_sub(end, start));
+		min_response_time = min(min_response_time, ipc_response_time);
+		max_response_time = max(max_response_time, ipc_response_time);
+
+		/* sum up response times */
+		avg_response_time += ipc_response_time;
+		i++;
+
+		/* test complete? */
+		if (flood_duration_test) {
+			if (ktime_to_ns(end) >= test_end)
+				break;
+		} else {
+			if (i == ipc_count)
+				break;
+		}
+	}
+
+	if (ret < 0)
+		dev_err(sdev->dev,
+			"error: ipc flood test failed at %d iterations\n", i);
+
+	/* compute average response time */
+	do_div(avg_response_time, i);
+
+	if (flood_duration_test) {
+		dev_dbg(sdev->dev, "IPC Flood test duration: %lldms\n",
+			(long long)ipc_duration_ms);
+		output1 = kasprintf(GFP_KERNEL, "IPC Flood duration: %lldms\n",
+				    (long long)ipc_duration_ms);
+		if (!output1)
+			return -ENOMEM;
+	}
+
+	/* clear previous test output */
+	memset(dfse->cache_buf, 0, IPC_FLOOD_TEST_RESULT_LEN);
+
+	dev_dbg(sdev->dev,
+		"IPC Flood count: %d, Avg response time: %lldns\n",
+		i, (long long)avg_response_time);
+	dev_dbg(sdev->dev, "Max response time: %lldns\n",
+		(long long)max_response_time);
+	dev_dbg(sdev->dev, "Min response time: %lldns\n",
+		(long long)min_response_time);
+
+	/* format test output */
+	output2 = kasprintf(GFP_KERNEL, "IPC flood count: %d\n", i);
+	if (!output2)
+		return -ENOMEM;
+
+	output3 = kasprintf(GFP_KERNEL, "Avg response time: %lldns\n",
+			    (long long)avg_response_time);
+	if (!output3)
+		return -ENOMEM;
+
+	output4 = kasprintf(GFP_KERNEL, "Max response time: %lldns\n",
+			    (long long)max_response_time);
+	if (!output4)
+		return -ENOMEM;
+
+	output5 = kasprintf(GFP_KERNEL, "Min response time: %lldns\n",
+			    (long long)min_response_time);
+	if (!output5)
+		return -ENOMEM;
```

Comment:
@ranj063 This is now very leaky. If output4 is succesfully allocated, but output5 fails, you'll leak memory for all allocs done before as you return directly here. 

Idea: It would seem easier that you only store the integer test data in this function, and do the pretty-printing to ascii in a separate function when the data is read. You can use snprintf() to directly output to a kmalloc'ed buffer to avoid the many outputX variables. You can call snprintf() once with a NULL buffer to figure out how much memory is needed.

Reply:
@kv2019i of course this sucks. Sorry for wasting your time. let me fix the memory leaks. And thanks for the suggestion about the output formatting. will work on it today.

Reply:
@ranj063 No worries, pretty-printing in kernel is bound to be tricky. So @plbossart was not keen on snprintf(NULL), hmm.

The many local strings still looks a bit odd. kfree(NULL) is valid, so how about if you do "goto out" after each kasprintf() failure and then kfree all output variables (and initialize them to NULL at start)? Now you return directly after output1/2 allocs, but use goto for output3.  I'll leave this as a comment, I'm ok to approve the patch in general -- I'll while for your response.

Reply:
@kv2019i either way looks odd. eventually i guess its up to the maintainers to decide what looks best. I could initialize all to null and free them all in case of error but I think it is less intuitive.